python-package/xgboost/distributed.py

```diff
@@ -0,0 +1,369 @@
+        self.n_cols = X.shape[1]
+
+        self.X = X
+        self.y = y
+        self.weights = weight
+
+        assert type(X) in (DDF, DA)
+        assert type(y) in (DDF, DS, DA)
+
+        client = get_client()
+
+        if isinstance(X, DDF):
+            client.sync(self.from_dataframe)
+        elif isinstance(X, DA):
+            client.sync(self.from_array)
+
+    async def from_dataframe(self):
+        '''Obtain references to dataframe partitions.'''
+        # Breaking data into partitions, a trick borrowed from dask_xgboost
+        client = get_client()
+        # to_delayed downgrades high-level objects into numpy or
+        # pandas equivalents.
+        X_parts = self.X.to_delayed()
+        y_parts = self.y.to_delayed()
+
+        assert len(X_parts) == len(
+            y_parts), 'Partitions between X and y are not consistent'
+
+        if self.weights:
+            w_parts = self.weights.to_delayed()
+            assert len(X_parts) == len(
+                w_parts), 'Partitions between X and weight are not consistent.'
+            parts = list(map(delayed, zip(X_parts, y_parts, w_parts)))
+        else:
+            parts = list(map(delayed, zip(X_parts, y_parts)))
+
+        parts = client.compute(parts)
+        await wait(parts)  # async wait for parts to be computed
+
+        for part in parts:
+            assert part.status == 'finished'
+
+        key_to_partition = dict([(part.key, part) for part in parts])
+        who_has = await client.scheduler.who_has(
+            keys=[part.key for part in parts])
+
+        worker_map = defaultdict(list)
+        for key, workers in who_has.items():
+            worker_map[first(workers)].append(key_to_partition[key])
+        self.worker_map = worker_map
+
+    async def from_array(self):
+        self.worker_map = await map_local_data(self.X, self.y, self.weights)
+
+    def get_worker_data(self, worker):
+        '''Get data that local to worker.
+        worker: The worker used as key to data.
+
+        Returns
+        -------
+        A DMatrix object.
+        '''
+        list_of_parts = self.worker_map[worker.address]
+        list_of_parts = [p.result() for p in list_of_parts]
+
+        if self.weights:
+            data, labels, weights = zip(*list_of_parts)
+        else:
+            data, labels = zip(*list_of_parts)
+
+        data = concat(data)
+        labels = concat(labels)
+        if self.weights:
+            weights = concat(weights)
+        else:
+            weights = None
+
+        dmatrix = DMatrix(data,
+                          labels,
+                          weight=weights,
+                          missing=self._missing,
+                          feature_names=self._feature_names,
+                          feature_types=self._feature_types)
+        return dmatrix
+
+    def num_row(self):
+        return self.n_rows
+
+    def num_col(self):
+        return self.n_cols
+
+
+def train(worker_train,
```

Comment:
Recommend adding a keyword argument for `client`, here. Like this:
```suggestion
def train(client=get_client(),
          worker_train,
```
This way, `get_client()` is the default client; otherwise, it can be overridden by the user.

Reply:
If you do this, then you can add a check `if client == None` based on previous import statements.

Reply:
@mt-jones Maybe I need to make this `train` method as another interface.